openedx/core/djangoapps/discussions/tasks.py

```diff
@@ -45,38 +45,54 @@ def update_discussions_settings_from_course(course_key: CourseKey) -> CourseDisc
     store = modulestore()
 
     supports_in_context = DiscussionsConfiguration.get(course_key).supports_in_context_discussions()
+    provider_type = DiscussionsConfiguration.get(course_key).provider_type
 
     def iter_discussable_units():
         sections = store.get_items(course_key, qualifiers={'category': 'sequential'})
+        # Start at -1 so that the initial increment starts it at 0.
+        idx = -1
         for section in sections:
             for unit in section.get_children():
+                # Increment index even for skipped units so that the index is more stable and won't change
+                # if settings change, only if a unit is added or removed.
+                idx += 1
                 # If unit-level visibility is enabled and the unit doesn't have discussion enabled, skip it.
                 if unit_level_visibility and not getattr(unit, 'discussion_enabled', False):
                     continue
                 # If the unit is in a graded section and graded sections aren't enabled skip it.
                 if section.graded and not enable_graded_units:
                     continue
+                # If the unit is an exam, skip it.
+                if section.is_practice_exam or section.is_proctored_enabled or section.is_time_limited:
+                    continue
                 yield DiscussionTopicContext(
                     usage_key=unit.location,
                     title=unit.display_name,
                     group_id=None,
+                    ordering=idx,
                 )
 
     with store.branch_setting(ModuleStoreEnum.Branch.published_only, course_key):
         course = store.get_course(course_key)
-        provider = course.discussions_settings.get('provider')
+        provider = course.discussions_settings.get('provider', provider_type)
         enable_in_context = course.discussions_settings.get('enable_in_context', True)
         provider_config = course.discussions_settings.get(provider, {})
         unit_level_visibility = course.discussions_settings.get('unit_level_visibility', False)
         enable_graded_units = course.discussions_settings.get('enable_graded_units', False)
         contexts = []
         if supports_in_context:
+            sorted_topics = sorted(
+                course.discussion_topics.items(),
+                key=lambda item: item[1].get("sort_key", item[0])
+            )
```

Comment:
Sort the topics by sort_key by default. since this is how it currently works.

Reply:
Can you explain what is "sort_key"? where the value is coming from?

Reply:
It doesn't seem to be a documented feature, but for each discussion topic in the course advanced settings, you can provide a sort key to allow sorting the topic if your own custom order. 

You can see it in use here for instance: lms/djangoapps/discussion/rest_api/api.py:397